{"cells":[{"cell_type":"markdown","metadata":{"id":"BUZH2533wuuq"},"source":["# **Introducción a Python**\n","# FP25. Manejo de errores y excepciones"]},{"cell_type":"markdown","metadata":{"id":"CXOHoI0rwuut"},"source":["¡Incluso los agentes especiales cometen errores! Definitivamente has cometido errores hasta este punto de tu entrenamiento. Veamos qué se produce cuando obtenemos un error e intentemos comprenderlos mejor:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9wsl8Ecrwuuu","outputId":"ad4221dd-0bd1-4e9e-b3e7-834f4ac7a7a1"},"outputs":[{"ename":"SyntaxError","evalue":"EOL while scanning string literal (2462765896.py, line 3)","output_type":"error","traceback":["\u001b[0;36m  File \u001b[0;32m\"/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/2462765896.py\"\u001b[0;36m, line \u001b[0;32m3\u001b[0m\n\u001b[0;31m    print('hola)\u001b[0m\n\u001b[0m                ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m EOL while scanning string literal\n"]}],"source":["# Ejecuta esta celda tal y como está ...\n","\n","print('hola)"]},{"cell_type":"markdown","metadata":{"id":"c4cwxjb9wuuv"},"source":["Fíjate que obtenemos un **SyntaxError**, con la descripción adicional de que fue un **EOL (Error de fin de línea) mientras se escanea la cadena literal**. Esto es lo suficientemente específico para que veamos que olvidamos una comilla simple al final de la línea. Comprender estos diversos tipos de errores te ayudarán a depurar tu código mucho más rápido.\n","\n","Este tipo de error y descripción se conoce como una **Excepción** (*Exception*). Incluso si una declaración o expresión es sintácticamente correcta, puede causar un error cuando se intenta ejecutarla. Los errores detectados durante la ejecución se denominan excepciones y no son necesariamente fatales.\n","\n","Puedes consultar la lista completa de [excepciones](https://docs.python.org/3/tutorial/errors.html) aquí. Ahora aprendamos a manejar errores y excepciones en nuestro propio código."]},{"cell_type":"markdown","metadata":{"id":"vLTI3Hcnwuuw"},"source":["## <font color='blue'>**La sentencia `try` `except`**</font>\n","\n","La terminología y la sintaxis básicas que se utilizan para manejar errores en Python son las declaraciones `try` y `except`. El código que puede hacer que ocurra una excepción se coloca en el bloque `try` y el manejo de la excepción se implementa en el bloque de código `except`. Nuevamente la indentación es clave en esta estructura. La forma de sintaxis es:\n","```python\n","     try:\n","        # Intentas tu operación aquí\n","     except ExceptionI:\n","        # Si hay ExceptionI, ejecuta este bloque.\n","     except ExceptionII:\n","        # Si hay ExceptionII, ejecuta este bloque.\n","     else:\n","        # Si no hay excepción, ejecuta este bloque.\n","     finally:\n","        # Siempre ejecuta este código\n","```\n","También podemos verificar cualquier excepción con solo usar `except`. Para comprender mejor todo esto, veremos un ejemplo de un código que abre y escribe en un archivo."]},{"cell_type":"markdown","metadata":{"id":"-NjkE3OAwuuw"},"source":["<font color='red'>Atención:</font>\n","\n","La siguiente línea de código creará un archivo en modo escritura en el directorio en el cual este notebook se está ejecutando."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"hoMxCJBiwuux","outputId":"08ced794-fd02-4330-ab6c-ebd198ca7fb5"},"outputs":[{"name":"stdout","output_type":"stream","text":["Contenido escrito exitosamente\n"]}],"source":["try:\n","    f = open('testfile','w')\n","    f.write('Prueba escribiendo esto')\n","\n","except IOError:\n","\n","    # Esto solo buscará una excepción del tipo IOError y luego ejecutará la declaración de print( )\n","    print(\"Error: Archivo no encontrado o no se pudo leer su data\")\n","else:\n","\n","    print(\"Contenido escrito exitosamente\")\n","    f.close()"]},{"cell_type":"markdown","metadata":{"id":"niDeY-Ttwuuy"},"source":["Ahora veamos qué pasaría si no tuviéramos permiso de escritura (abriendo solo con 'r'):"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"5eam3xn-wuuy","outputId":"59372e3b-6a15-4299-dd17-6ec65993927f"},"outputs":[{"name":"stdout","output_type":"stream","text":["Error: Archivo no encontrado o no se pudo leer su data\n"]}],"source":["try:\n","    f = open('testfile','r')\n","    f.write('Prueba escribiendo esto')\n","\n","except IOError:\n","\n","    # Esto solo buscará una excepción del tipo IOError y luego ejecutará la declaración de print( )\n","    print(\"Error: Archivo no encontrado o no se pudo leer su data\")\n","else:\n","\n","    print(\"Contenido escrito exitosamente\")\n","    f.close()"]},{"cell_type":"markdown","metadata":{"id":"lh_uCCzAwuuz"},"source":["Si queremos verificar varios errores, podemos usar sólo `except:`"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"uTkb6VRWwuuz","outputId":"49b9d80a-9713-4b4a-b80a-d8ec98660138"},"outputs":[{"name":"stdout","output_type":"stream","text":["Error: Archivo no encontrado o no se pudo leer su data\n"]}],"source":["try:\n","    f = open('testfile','r')\n","    f.write('Prueba escribiendo esto')\n","\n","except:\n","\n","    # En este caso se evaluarán todas las posibles excepciones\n","    print(\"Error: Archivo no encontrado o no se pudo leer su data\")\n","else:\n","\n","    print(\"Contenido escrito exitosamente\")\n","    f.close()"]},{"cell_type":"markdown","metadata":{"id":"S7BzbrBFwuuz"},"source":["De esta manera, no tendrás que preocuparte por memorizar todos los tipos de excepción posibles.\n","\n","## <font color='blue'>**La cláusula `finally`**</font>\n","Veamos ahora la palabra clave `finally`:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DCtLOiAuwuu0","outputId":"9aa8bcd2-f83e-4c53-a6bf-746999f1ca69"},"outputs":[{"name":"stdout","output_type":"stream","text":["Siempre se ejecuta el set de instruciones anidadas en 'finally'\n"]}],"source":["try:\n","    f = open(\"testfile\", \"w\")\n","    f.write(\"Prueba escribiendo esto\")\n","finally:\n","    print(\"Siempre se ejecuta el set de instrucciones anidadas en 'finally'\")\n","    f.close()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yeOaMz8zwuu0","outputId":"db52003f-c4cb-4eb6-ac9d-cec11ccee795"},"outputs":[{"name":"stdout","output_type":"stream","text":["Siempre se ejecuta el set de instruciones anidadas en 'finally'\n"]},{"ename":"UnsupportedOperation","evalue":"not writable","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mUnsupportedOperation\u001b[0m                      Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/4120907272.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"testfile\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Prueba escribiendo esto\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Siempre se ejecuta el set de instruciones anidadas en 'finally'\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mUnsupportedOperation\u001b[0m: not writable"]}],"source":["try:\n","\n","    f = open(\"testfile\", \"r\")\n","    f.write(\"Prueba escribiendo esto\")\n","finally:\n","    print(\"Siempre se ejecuta el set de instrucciones anidadas en 'finally'\")"]},{"cell_type":"markdown","metadata":{"id":"6nq40ZDcwuu0"},"source":["Podemos usar esto junto con la palabra clave `except`; con ello obtenemos una estructura de manejo de errores del tipo: `try`, `except`, `else` y `finally`."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"d3Gjq2IZwuu0","outputId":"d11b519b-0e6e-46c2-bba9-b31e5b648367"},"outputs":[{"name":"stdout","output_type":"stream","text":["Error: Archivo no encontrado o no se pudo leer su data\n","Siempre se ejecuta el set de instruciones anidadas en \"finally\"\n"]}],"source":["try:\n","    f = open('testfile','r')\n","    f.write('Prueba escribiendo esto')\n","\n","except:\n","\n","    # En este caso se evaluarán todas las posibles excepciones\n","    print(\"Error: Archivo no encontrado o no se pudo leer su data\")\n","else:\n","\n","    print(\"Contenido escrito exitosamente\")\n","\n","finally:\n","    f.close()\n","    print('Siempre se ejecuta el set de instrucciones anidadas en \"finally\"')"]},{"cell_type":"markdown","metadata":{"id":"SSL9PW0Ywuu1"},"source":["## <font color='blue'>**La palabra clave `raise`**</font>\n","En Python, se generan excepciones cuando se producen errores en tiempo de ejecución. También podemos generar excepciones manualmente usando la palabra clave `raise`.\n","\n","Opcionalmente, podemos pasar valores a la excepción para aclarar por qué se generó esa excepción.\n","\n","La sintáxis completa es la siguiente:\n","```python\n","raise NombreError('Mensaje de error (opcional)')  # Con mensaje de error\n","\n","raise NombreError                                 # Si mensaje de error\n","```\n","Veamos un par de ejemplos"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jXVnNxBrwuu1","outputId":"d06ea191-9081-44d7-b9a5-0909f9997a89"},"outputs":[{"ename":"Exception","evalue":"Sólo números mayores a cero","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/246459123.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mx\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Sólo números mayores a cero'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mException\u001b[0m: Sólo números mayores a cero"]}],"source":["# Genera una excepción si el valor es menor a cero\n","x = -1\n","if x <= 0:\n","    raise Exception('Sólo números mayores a cero')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"6xpknXTgwuu1"},"outputs":[],"source":["# Ingresar un año\n","\n","def isBisiesto(year):\n","    if year.isnumeric():\n","        print('Calcula año bisiento')\n","    else:\n","        raise TypeError('Solo ingresar números enteros')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"6aRF58puwuu1","outputId":"4a6eb818-4bcf-4212-ba1b-fe2b56eba4d9"},"outputs":[{"name":"stdout","output_type":"stream","text":["Calcula año bisiento\n"]}],"source":["isBisiesto('2021')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MRpP8b1lwuu2","outputId":"e06008a3-ea70-44b7-a12d-ae8f8b95de43"},"outputs":[{"ename":"TypeError","evalue":"Solo ingresar números enteros","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/2641102445.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0misBisiesto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'hola'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/4148583260.py\u001b[0m in \u001b[0;36misBisiesto\u001b[0;34m(year)\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Calcula año bisiento'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Solo ingresar números enteros'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mTypeError\u001b[0m: Solo ingresar números enteros"]}],"source":["isBisiesto('hola')"]},{"cell_type":"markdown","metadata":{"id":"pWmtRKHJwuu2"},"source":["Veamos lo mismo con `try`, `except`"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"C0HjIsolwuu2"},"outputs":[],"source":["def isBisiesto2(year):\n","    try:\n","        int(year)\n","\n","    except ValueError:\n","        raise TypeError('Solo ingresar números enteros')\n","\n","    else:\n","        print('Calcula año bisiento')\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"FlNNBZ8Wwuu2","outputId":"5c7951ef-92e0-4657-d5d1-41e33726d4f5"},"outputs":[{"name":"stdout","output_type":"stream","text":["Calcula año bisiento\n"]}],"source":["isBisiesto2('1998')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"TLTn9jWZwuu2","outputId":"0075881a-87ac-49cb-e925-fa908db2409b"},"outputs":[{"ename":"TypeError","evalue":"Solo ingresar números enteros","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/2013402300.py\u001b[0m in \u001b[0;36misBisiesto2\u001b[0;34m(year)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m         \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0myear\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'hola'","\nDuring handling of the above exception, another exception occurred:\n","\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/4025596551.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0misBisiesto2\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'hola'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/2013402300.py\u001b[0m in \u001b[0;36misBisiesto2\u001b[0;34m(year)\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Solo ingresar números enteros'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mTypeError\u001b[0m: Solo ingresar números enteros"]}],"source":["isBisiesto2('hola')"]},{"cell_type":"markdown","source":["## <font color='blue'>__Ejercicios__</font>"],"metadata":{"id":"pXeh9WolSLIA"}},{"cell_type":"markdown","metadata":{"id":"56KW-_sewuu2"},"source":["### <font color='green'>Actividad 1: Challenging</font>\n","### Mejora tu función para el algoritmo del año bisiesto del notebook FP18\n","\n","Toma función que realizaste en el notebook FP18 y mejórala con lo siguiente:\n","\n","* Utiliza `try`, `except`, `else` para controlar errores\n","* Utiliza `raise` para generar excepciones\n","* Prueba isBisiesto3()\n","\n","Nombra tu función **isBisiesto3()**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DjditIq3wuu2"},"outputs":[],"source":["# Tu código aquí ...\n","def isBisiesto3(year):\n","    \"\"\"\n","    Verifica si el año ingresado es bisiesto o no\n","\n","    Parámetros:\n","    year (int): año a evaluar\n","\n","    Salida:\n","    (boolean): dependiendo si es o no bisiesto\n","\n","    \"\"\"\n","\n","    try:\n","        year = int(year)\n","\n","    except ValueError:\n","        raise TypeError('Solo ingresar números enteros')\n","\n","    else:\n","        if (year % 4 == 0) and (not (year % 100 == 0) or (year % 400 == 0)):\n","            return True\n","        else:\n","            return False\n","\n",""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"W1WxfZziwuu3","outputId":"21686766-99db-4f1e-e90f-d4217c60959c"},"outputs":[{"data":{"text/plain":["False"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["isBisiesto3('2018')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"F8Od1tE_wuu3","outputId":"85beb7c0-61a5-4914-a7ed-990117eba127"},"outputs":[{"data":{"text/plain":["True"]},"execution_count":17,"metadata":{},"output_type":"execute_result"}],"source":["isBisiesto3('2020')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"2wlldjiDwuu3","outputId":"d8ac3f16-dcab-4737-e6bc-a4cb8d982645"},"outputs":[{"ename":"TypeError","evalue":"Solo ingresar números enteros","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/1228210317.py\u001b[0m in \u001b[0;36misBisiesto3\u001b[0;34m(year)\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m         \u001b[0myear\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0myear\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'hola'","\nDuring handling of the above exception, another exception occurred:\n","\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/1325362559.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0misBisiesto3\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'hola'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/1228210317.py\u001b[0m in \u001b[0;36misBisiesto3\u001b[0;34m(year)\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Solo ingresar números enteros'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mTypeError\u001b[0m: Solo ingresar números enteros"]}],"source":["isBisiesto3('hola')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ZShQq7e8wuu3","outputId":"d44b7389-e2ff-4a6f-b651-935e87d57a05"},"outputs":[{"ename":"TypeError","evalue":"Solo ingresar números enteros","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/1228210317.py\u001b[0m in \u001b[0;36misBisiesto3\u001b[0;34m(year)\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m         \u001b[0myear\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0myear\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '2018.0'","\nDuring handling of the above exception, another exception occurred:\n","\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/1120357096.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0misBisiesto3\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'2018.0'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/var/folders/2k/stjkgscn4v5bs3y7nm50vtb40000gn/T/ipykernel_91124/1228210317.py\u001b[0m in \u001b[0;36misBisiesto3\u001b[0;34m(year)\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Solo ingresar números enteros'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mTypeError\u001b[0m: Solo ingresar números enteros"]}],"source":["isBisiesto3('2018.0')"]},{"cell_type":"markdown","metadata":{"id":"frQjbvRCwuu3"},"source":["<font color='green'>Fin actividad 1</font>"]},{"cell_type":"markdown","source":["<img src=\"https://drive.google.com/uc?export=view&id=1DNuGbS1i-9it4Nyr3ZMncQz9cRhs2eJr\" width=\"100\" align=\"left\" title=\"Runa-perth\">\n","<br clear=\"left\">\n","\n","## <font color='blue'>**Resumen**</font>\n","La cláusula `try` en Python se utiliza para atrapar y manejar excepciones. Las excepciones son errores que ocurren durante la ejecución del programa. Cuando ocurre una excepción en el código, el flujo normal del programa se interrumpe. Si la excepción no se maneja, el programa se detendrá y mostrará un mensaje de error.\n","\n","Aquí es donde entra en juego la cláusula `try`. Puedes poner el código que podría generar una excepción dentro de un bloque `try`. Si ocurre una excepción en el bloque try, el flujo del programa se pasa a un bloque `except` correspondiente, donde se puede manejar la excepción.\n","\n","Además, la declaración `try` puede tener una cláusula `else`. El código dentro del bloque `else` se ejecuta si el código dentro del bloque `try` no generó ninguna excepción.\n","\n","Finalmente, la cláusula `finally` es opcional y se puede agregar después de los bloques `try` y `except`. El código dentro del bloque `finally` se ejecuta sin importar si se produjo una excepción o no. Es útil para acciones de limpieza que deben ser ejecutadas independientemente de si ocurrió un error.\n","\n","Aquí tienes un ejemplo:\n","\n","```python\n","try:\n","    # Código que podría generar una excepción\n","    resultado = 10 / 0\n","except ZeroDivisionError:\n","    # Esto se ejecuta si hay una excepción ZeroDivisionError\n","    print(\"Ha ocurrido una división por cero.\")\n","else:\n","    # Esto se ejecuta si no hay excepciones\n","    print(\"La división se realizó con éxito.\")\n","finally:\n","    # Esto se ejecuta sin importar si hay una excepción o no\n","    print(\"Esto se imprime siempre.\")\n","\n","```\n","En este ejemplo, la división por cero genera una excepción `ZeroDivisionError`, por lo que el bloque `except` se ejecuta. Luego, sin importar si se produjo una excepción o no, el bloque `finally` también se ejecuta.\n","\n","<img src=\"https://drive.google.com/uc?export=view&id=1DNuGbS1i-9it4Nyr3ZMncQz9cRhs2eJr\" width=\"50\" align=\"left\" title=\"Runa-perth\">\n","<br clear=\"left\">"],"metadata":{"id":"5J_-4m0NSzBl"}},{"cell_type":"markdown","metadata":{"id":"YqhW9kDJwuu3"},"source":["Muy bien hecho!!"]}],"metadata":{"anaconda-cloud":{},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.11"},"colab":{"provenance":[],"toc_visible":true}},"nbformat":4,"nbformat_minor":0}